python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll

```diff
@@ -0,0 +1,1540 @@
+    result = DataFlowFunction.super.getParameter(ppos)
+  }
+}
+
+/**
+ * A module. This is not actually a callable, but we need this so a
+ * `ModuleVariableNode` have an enclosing callable.
+ */
+class DataFlowModuleScope extends DataFlowCallable, TModule {
+  Module mod;
+
+  DataFlowModuleScope() { this = TModule(mod) }
+
+  override string toString() { result = mod.toString() }
+
+  override string getQualifiedName() { result = mod.getName() }
+
+  override Module getScope() { result = mod }
+
+  override Location getLocation() { result = mod.getLocation() }
+
+  override ParameterNode getParameter(ParameterPosition ppos) { none() }
+}
+
+class LibraryCallableValue extends DataFlowCallable, TLibraryCallable {
+  LibraryCallable callable;
+
+  LibraryCallableValue() { this = TLibraryCallable(callable) }
+
+  override string toString() { result = "LibraryCallableValue: " + callable.toString() }
+
+  override string getQualifiedName() { result = callable.toString() }
+
+  /** Gets a data-flow node, where this library callable is used as a call-back. */
+  ArgumentNode getACallback() { result = callable.getACallback() }
+
+  override Scope getScope() { none() }
+
+  override ParameterNode getParameter(ParameterPosition ppos) { none() }
+
+  override LibraryCallable asLibraryCallable() { result = callable }
+
+  override Location getLocation() { none() }
+}
+
+// =============================================================================
+// Type trackers used to resolve calls.
+// =============================================================================
+/** Gets a call to `type`. */
+private CallCfgNode getTypeCall() {
+  exists(NameNode id | id.getId() = "type" and id.isGlobal() |
+    result.getFunction().asCfgNode() = id
+  )
+}
+
+/** Gets a call to `super`. */
+private CallCfgNode getSuperCall() {
+  // While it is possible to reference super and call it later, it's almost never done in
+  // practice. From looking at top 1000 projects, there were a few uses around mocking (see
+  // link below), but otherwise only 2 edgecases. Overall it seems ok to ignore this complexity.
+  //
+  // https://github.com/python/cpython/blob/18b1782192f85bd26db89f5bc850f8bee4247c1a/Lib/unittest/mock.py#L48-L50
+  exists(NameNode id | id.getId() = "super" and id.isGlobal() |
+    result.getFunction().asCfgNode() = id
+  )
+}
+
+/**
+ * Gets a reference to the function `func`.
+ */
+private TypeTrackingNode functionTracker(TypeTracker t, Function func) {
+  t.start() and
+  (
+    result.asExpr() = func.getDefinition()
+    or
+    // when a function is decorated, it's the result of the (last) decorator call that
+    // is used
+    result.asExpr() = func.getDefinition().(FunctionExpr).getADecoratorCall()
+  )
+  or
+  exists(TypeTracker t2 | result = functionTracker(t2, func).track(t2, t))
+}
+
+/**
+ * Gets a reference to the function `func`.
+ */
+Node functionTracker(Function func) { functionTracker(TypeTracker::end(), func).flowsTo(result) }
```

Comment:
I guess it would "only" take adding the `flowsTo` to the appropriate places. Do you expect this to pay off in terms of performance, or is it rather because it would just be a nice way to do things?

Reply:
It's just a thing that's been continually bothering me -- the fact that unlike JavaScript (where `SourceNode`s are tied into everything), we have to write these extra `flowsTo` bits all over our type trackers.

I don't think it affects performance much, though, unless you have a truly enormous number of nodes (and in that case you probably have bigger issues to contend with).